File: Assets/src/model/tasks/ChaseTargetTask.cs
using UnityEngine;
/// Open-ended.
[System.Serializable]
public class ChaseTargetTask : MoveNextToTargetTask {
  public override TaskStage WhenToCheckIsDone => TaskStage.Before;
  protected Body targetBody;
  private int extraMovesCutoff;
  public override Vector2Int target => targetBody.pos;

  public ChaseTargetTask(Actor actor, Body targetBody, int extraMovesCutoff = 0) : base(actor, targetBody.pos) {
    this.targetBody = targetBody;
    this.extraMovesCutoff = extraMovesCutoff;
  }

  public Body GetTargetBody() {
    return targetBody;
  }

  public override void PreStep() {
    if (targetBody == null || targetBody is Player p && p.isCamouflaged) {
      this.path.Clear();
    } else {
      this.path = FindBestAdjacentPath(actor, target);
      if (extraMovesCutoff > 0 && this.path.Count >= extraMovesCutoff) {
        this.path.RemoveRange(this.path.Count - extraMovesCutoff, extraMovesCutoff);
      }
    }
  }
}

File: web/src/model/tasks/ChaseTargetTask.ts
import { MoveNextToTargetTask } from './MoveNextToTargetTask';
import { TaskStage } from '../ActorTask';
import { Vector2Int } from '../../core/Vector2Int';
import type { Actor } from '../Actor';
import type { Body } from '../Body';
import type { BaseAction } from '../BaseAction';

export class ChaseTargetTask extends MoveNextToTargetTask {
  protected targetBody: Body;
  private extraMovesCutoff: number;

  get whenToCheckIsDone(): TaskStage {
    return TaskStage.Before;
  }

  get target(): Vector2Int {
    return this.targetBody.pos;
  }

  constructor(actor: Actor, targetBody: Body, extraMovesCutoff = 0) {
    super(actor, targetBody.pos);
    this.targetBody = targetBody;
    this.extraMovesCutoff = extraMovesCutoff;
  }

  getTargetBody(): Body {
    return this.targetBody;
  }

  preStep(): void {
    if (!this.targetBody || this.targetBody.isDead) {
      this.path = [];
    } else {
      this.path = MoveNextToTargetTask.findBestAdjacentPath(this.actor, this.targetBody.pos);
      if (this.extraMovesCutoff > 0 && this.path.length >= this.extraMovesCutoff) {
        this.path.splice(this.path.length - this.extraMovesCutoff, this.extraMovesCutoff);
      }
    }
  }
}
